### PR TITLE
Add bounds checking to wctob for invalid wide characters

### LINKED ISSUE #1355: picolibc's `wctob()` accepts invalid negative values as one-byte characters
According to cppreference.com,

[`wctob()`](https://cppreference.com/w/c/string/multibyte/wctob) must return `EOF` if its argument does not represent a multibyte character with length one in the initial shift state.

picolibc currently rejects `WEOF`, but passes other values directly to the active wide-to-multibyte converter:
```c
if (wc == WEOF)
    return EOF;

return __WCTOMB((char *)pmb, wc, &mbs) == 1
           ? (int)pmb[0]
           : EOF;
```
Consequently, on targets where wchar_t is signed, negative values other than WEOF are incorrectly reported as successful one-byte conversions.

For example, the ASCII conversion accepts any value less than `0x80`:

```c
if (wchar >= 0x80)
    return -1;

*s = (char)wchar;
return 1;
```

The UTF-8 conversion has similar behavior:

```c
if (wchar <= 0x7f) {
    *s = wchar;
    return 1;
}
```

As a result, `-2` and `-3` are converted to `unsigned char` values instead of being rejected.

## Reproducer:

```c
#include <stdio.h>
#include <wchar.h>

int
main(void)
{
    int ret1 = wctob(-1);
    int ret2 = wctob(-2);
    int ret3 = wctob(-3);
    int pass = ret1 == EOF && ret2 == EOF && ret3 == EOF;

    printf("wctob(-1)=%d wctob(-2)=%d wctob(-3)=%d "
           "(want EOF=%d) %s\n",
           ret1, ret2, ret3, EOF, pass ? "PASS" : "FAIL");

    return pass ? 0 : 1;
}
```

```bash
$ clang \
  --target=hexagon-unknown-none-elf \
  -T picolibc.ld \
  --cstdlib=picolibc \
  wctob-negative-input.c \
  -o wctob-negative-input.pico

$ qemu-system-hexagon \
  -cpu v68 \
  -M virt \
  -kernel ./wctob-negative-input.pico
```

Output:

```text
wctob(-1)=-1 wctob(-2)=254 wctob(-3)=253 (want EOF=-1) FAIL
```

For comparison, glibc returns `EOF` for all three calls.

### PR DESCRIPTION
**Pull request by SuperChamp234**
Add stdint.h include and validate that wc doesn't exceed WCHAR_MAX before processing. Add test cases for negative wint_t values.

Tries to fix #1355

### CODE REVIEW

**Comment by SuperChamp234** [comment]
Ah let me figure something out for arm

File: libc/stdlib/wctob.c [hunk 1]
```diff
@@ -13,7 +14,8 @@ wctob(wint_t wc)
     mbstate_t     mbs;
     unsigned char pmb[MB_LEN_MAX];
 
-    if (wc == WEOF)
+    /* Reject sign or upper bits outside the range of wchar_t. */
+    if (wc == WEOF || ((uintmax_t)wc & ~(uintmax_t)WCHAR_MAX) != 0)
```
Comment (keith-packard) [inline libc/stdlib/wctob.c]: I can't find any wording in the C standard which requires `WCHAR_MAX` to be of the form `2^n-1`, so we can't use a bit-wise comparison here. Why does the simple (uintmax_t)wc > (uintmax_t)WCHAR_MAX not work here?

Comment (SuperChamp234) [reply libc/stdlib/wctob.c]: It fails for arm which uses unsigned int for `wchar_t` :(

**Review by keith-packard — APPROVED** [review]
Thanks for the fixup; I note that using an intermediate variable appears to keep the compiler from complaining about the 'this will always be true' while still letting it elide the test from the generated binary (which surprised me). Nicely done.
